### PR TITLE
CPP: Add query for CWE-805: Buffer Access with Incorrect Length Value using some functions

### PR DESCRIPTION
**Pull request by ihsinme**
This query is looking for a simple error condition in the argument. It seemed to me that in working with this problem, the functions of working with ssl were undeservedly forgotten.

It is worth noting that this request also has the potential to be expanded by simply adding new features.

[CVE-2004-2003](https://nvd.nist.gov/vuln/detail/CVE-2004-2003)

### CODE REVIEW

**Comment by MathiasVP** [comment]
Hi @ihsinme,

Thanks for another contribution! Seeing as we still have [quite a backlog](https://github.com/github/codeql/pulls/ihsinme) to review from you it'll probably take some time before we get to it.

**Review by jketema — COMMENTED** [review]
Thanks for the contribution. How does this query relate to `cpp/overrunning-write` and the `BufferWrite` library?

**Comment by ihsinme** [comment]
> Thanks for the contribution. How does this query relate to `cpp/overrunning-write` and the `BufferWrite` library?

if you mean how it differs then it is:
1. functions set
2. simplicity , I only consider the situation `sizeof` to `const`

I noticed that some functions overlap (((   I remove them.

**Comment by ihsinme** [comment]
I think you know better.

but my query is just expanding, it actually needs a mechanism to look up the functions contained in the main call `sizeof(var)` and `var`, and added the name of that function and the argument number.

File: cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.qhelp [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>Using a size argument that is larger than the buffer size will result in an out-of-memory and possibly overflow. You need to limit the value of the length argument.</p>
```
Comment (jketema) [inline cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.qhelp:6]: What do you mean by "out-of-memory"? Do you mean an "out-of-bounds memory access"?

**Review by jketema — COMMENTED** [review]
Besides the qhelp problem this looks ok. As I mentioned in the other PR I'm reviewing, more descriptive variables names would help, `b`, `s`, `bpos` and `spos` are not very descriptive. I can live with `v` and `at` as those are very local scope in the query, and cannot be confused with anything else.

Using some `class`-based approach might also have been useful here, especially to group names that logically belong to the same library `X509`, `SSL`, `BIO`, `EVP`, `BN`, ...

Were you planning to submit this for a security bounty?

File: cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.ql [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ * @id cpp/buffer-access-with-incorrect-length-value
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-805
+ */
+
+import cpp
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
```
Comment (jketema) [inline cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.ql:14]: I don't think GVN is being used?

File: cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.qhelp (hunk 1 above)
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.qhelp]: Thanks

File: cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.ql (hunk 2 above)
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.ql]: Thanks

**Comment by ihsinme** [comment]
> Besides the qhelp problem this looks ok. As I mentioned in the other PR I'm reviewing, more descriptive variables names would help, `b`, `s`, `bpos` and `spos` are not very descriptive. I can live with `v` and `at` as those are very local scope in the query, and cannot be confused with anything else.

you're right. what do you think about:
```ql
  exists(ArrayType array, int bpos, int spos |
    numberArgument(fc.getTarget(), bpos, spos) and
    fc.getArgument(spos).getValue().toInt() > array.getByteSize() and
    fc.getArgument(bpos).(VariableAccess).getTarget().getADeclarationEntry().getType() = array
  )
```

> Using some `class`-based approach might also have been useful here, especially to group names that logically belong to the same library `X509`, `SSL`, `BIO`, `EVP`, `BN`, ...

I thought about it, especially when the first big group was formed to include argument positions. but then he abandoned this idea, since many groups containing one or two functions appeared.
if you still think so, I'll probably move the names to classes and add `instanceof` to the classes.

> Were you planning to submit this for a security bounty?

yes I plan

**Comment by jketema** [comment]
> you're right. what do you think about:

Not very helpful, as I need to dig into the defined predicate to figure out what `bpos` and `spos` are. More helpful would be to name those `bufferArgumentPosition` and `sizeArgumentPosition` (or `bufArgPos` and `sizeArgPos`, or something like that).

**Comment by jketema** [comment]
> Do I need to do this for this PR only, or do I have to generate pre-merge requests for all PRs?

Preferably for all.

File: cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.ql [hunk 3]
```diff
@@ -0,0 +1,71 @@
+/**
+ * @name Buffer access with incorrect length value.
```
Comment (jketema) [inline cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.ql:2]: Apologies for missing this earlier.
```suggestion
 * @name Buffer access with incorrect length value
```

File: cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.ql [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  or
+  f.hasGlobalOrStdName("getnameinfo") and bpos = 4 and spos = 5
+}
+
+from FunctionCall fc
+where
+  exists(ArrayType array, int bufArgPos, int sizeArgPos |
+    numberArgument(fc.getTarget(), bufArgPos, sizeArgPos) and
+    fc.getArgument(sizeArgPos).getValue().toInt() > array.getByteSize() and
+    fc.getArgument(bufArgPos).(VariableAccess).getTarget().getADeclarationEntry().getType() = array
```
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.ql:68]: This showed up on some of the larger databases we test on. This is also clearly not a fix you could have come up with by yourself.

[the remark this answers is not in the document]
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.ql]: @jketema, thanks a lot.
now I don't even know how to measure the slowdown in individual commands, if you could tell me where to go in this matter, in subsequent queries I would be more helpful.

thanks again.

[the remark this answers is not in the document]
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-805/BufferAccessWithIncorrectLengthValue.ql]: thanks.
I will try to use this in the future.

**Comment by jketema** [comment]
@ihsinme The formatting of the query file doesn't follow our guidelines, could you please reformat? Thanks.

**Comment by ihsinme** [comment]
good afternoon @jketema.
any news on this PR?

**Comment by jketema** [comment]
Please ping security lab about this. Note that people may be on holiday, so response might be slow.

**Review by jketema — APPROVED** [review]
Approved. Apologies that this took so long. Thanks for your contribution.

**Pull request merged**